akka-cluster-sharding-typed/src/test/scala/akka/cluster/sharding/typed/scaladsl/ClusterShardingPersistenceSpec.scala

```diff
@@ -33,69 +45,350 @@ object ClusterShardingPersistenceSpec {
 
-      val ref = ClusterSharding(system).entityRefFor(typeKey, "456")
+      val ref = ClusterSharding(system).entityRefFor(typeKey, entityId)
       val done1 = ref ? AddWithConfirmation("a")
       done1.futureValue should ===(Done)
 
       val done2: Future[Done] = ref ? AddWithConfirmation("b")
       done2.futureValue should ===(Done)
 
       ref ! Get(p.ref)
-      p.expectMessage("456:a|b")
+      p.expectMessage("2:a|b")
+    }
+
+    "handle PoisionPill after persist effect" in {
+      val entityId = nextEntityId()
+      val recoveryProbe = TestProbe[String]()
+      recoveryCompletedProbes.put(entityId, recoveryProbe.ref)
+
+      val p1 = TestProbe[Done]()
+      val ref = ClusterSharding(system).entityRefFor(typeKey, entityId)
+      (1 to 10).foreach { n ⇒
+        ref ! PassivateAndPersist(n.toString)(p1.ref)
+        // recoveryCompleted each time, verifies that it was actually stopped
+        recoveryProbe.expectMessage("recoveryCompleted:" + (1 until n).map(_.toString).mkString("|"))
+        p1.expectMessage(Done)
+      }
+
+      val p2 = TestProbe[String]()
+      ref ! Get(p2.ref)
+      p2.expectMessage(entityId + ":" + (1 to 10).map(_.toString).mkString("|"))
+    }
+
+    "handle PoisionPill after several stashed commands that persist" in {
+      val entityId = nextEntityId()
+      val actorRefPromise = Promise[ActorRef[Any]]()
+      entityActorRefs.put(entityId, actorRefPromise)
+
+      val addProbe = TestProbe[Done]()
+      val recoveryProbe = TestProbe[String]()
+      recoveryCompletedProbes.put(entityId, recoveryProbe.ref)
+
+      val entityRef = ClusterSharding(system).entityRefFor(typeKey, entityId)
+      // this will wakup the entity, and complete the entityActorRefPromise
+      entityRef ! AddWithConfirmation("a")(addProbe.ref)
+      addProbe.expectMessage(Done)
+      recoveryProbe.expectMessage("recoveryCompleted:")
+      // now we know that it's in EventSourcedRunning with no stashed commands
+
+      val actorRef = actorRefPromise.future.futureValue
+      // not sending via the EntityRef because that would make the test racy
+      // these are stashed, since before the PoisonPill
+      val latch = new CountDownLatch(1)
+      actorRef ! Block(latch)
+      actorRef ! AddWithConfirmation("b")(addProbe.ref)
+      actorRef ! AddWithConfirmation("c")(addProbe.ref)
+
+      actorRef ! PoisonPill
+
+      // those messages should be ignored since they happen after the PoisonPill,
+      actorRef ! AddWithConfirmation("d")(addProbe.ref)
+      actorRef ! AddWithConfirmation("e")(addProbe.ref)
+
+      // now we have enqueued the message sequence and start processing them
+      latch.countDown()
+
+      addProbe.expectMessage(Done)
+      addProbe.expectMessage(Done)
+
+      // wake up again
+      val p2 = TestProbe[String]()
+      p2.expectTerminated(actorRef, p2.remainingOrDefault)
+      Thread.sleep(500) // possible race that the Shard doesn't get Terminated first
```

Comment:
could we make this deterministic with an awaitAssert with a GetShardRegionState?

Reply:
good idea, done